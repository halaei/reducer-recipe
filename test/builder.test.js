import builder from '..';

const Artist = (r) => {
    const unknown = { name: 'Unknown', followers: 0, albums: [] };
    r.default(unknown)
        .on('set-artist-name', (state, action) => {
            return { ...state, name: action.name };
        })
        .on('set-artist-followers', (state, action) => {
            return { ...state, followers: action.followers };
        })
        .on('publish-new-album', (state, action) => {
            return { ...state, albums: [...state.albums, action.album] };
        });
};

const Album = (r) => {
    const unknown = {
        name: 'Unknown',
        artist: 'Unknown',
        cover: null,
        playing: false,
    };
    r.default(unknown)
        .on('set-album', (state, action) => {
            return { ...action.album, playing: false };
        })
        .on(['play-album', 'stop-album'], (state) => {
            return { ...state, playing: !state.playing };
        });
};

const User = (r) => {
    r.default(null)
        .on('login', (state, action) => {
            return action.username;
        });
};

const News = (state = [], action) => {
    switch (action.type) {
        case 'add-news':
            return [action.news, ...state];
        case 'pop-news':
            let newState = [...state];
            newState.pop();
            return newState;
        default:
            return state;
    }
}

const appReducer = builder.buildCombined(
    {
        App: builder.combine({ Album, Artist}),
        User,
    },
    {
        News,
    }
);

const reducer = (state, action) => {
    if (action.type === 'clear-store') {
        state = undefined;
    }
    return appReducer(state, action);
}

it('should generate the expected states', function () {
    let state = reducer(undefined, {type: undefined});
    expect(state).toMatchSnapshot();
    state = reducer(state, { type: 'login', username: 'hamid'});
    expect(state).toMatchSnapshot();
    state = reducer(state, { type: 'set-artist-name', name: 'Foo' });
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'set-album', album: {name: 'new', artist: 'Foo', cover: 'cover.jpg'}});
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'play-album'})
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'add-news', news: 'First news'});
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'add-news', news: 'Second news'});
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'add-news', news: 'Third news'});
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'pop-news'});
    expect(state).toMatchSnapshot();
    state = reducer(state, {type: 'clear-store'});
    expect(state).toMatchSnapshot();
});
