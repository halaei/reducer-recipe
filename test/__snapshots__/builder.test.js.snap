// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate the expected states 1`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Unknown",
      "cover": null,
      "name": "Unknown",
      "playing": false,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Unknown",
    },
  },
  "News": Array [],
  "User": null,
}
`;

exports[`should generate the expected states 2`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Unknown",
      "cover": null,
      "name": "Unknown",
      "playing": false,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Unknown",
    },
  },
  "News": Array [],
  "User": "hamid",
}
`;

exports[`should generate the expected states 3`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Unknown",
      "cover": null,
      "name": "Unknown",
      "playing": false,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [],
  "User": "hamid",
}
`;

exports[`should generate the expected states 4`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": false,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [],
  "User": "hamid",
}
`;

exports[`should generate the expected states 5`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": true,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [],
  "User": "hamid",
}
`;

exports[`should generate the expected states 6`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": true,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [
    "First news",
  ],
  "User": "hamid",
}
`;

exports[`should generate the expected states 7`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": true,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [
    "Second news",
    "First news",
  ],
  "User": "hamid",
}
`;

exports[`should generate the expected states 8`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": true,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [
    "Third news",
    "Second news",
    "First news",
  ],
  "User": "hamid",
}
`;

exports[`should generate the expected states 9`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Foo",
      "cover": "cover.jpg",
      "name": "new",
      "playing": true,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Foo",
    },
  },
  "News": Array [
    "Third news",
    "Second news",
  ],
  "User": "hamid",
}
`;

exports[`should generate the expected states 10`] = `
Object {
  "App": Object {
    "Album": Object {
      "artist": "Unknown",
      "cover": null,
      "name": "Unknown",
      "playing": false,
    },
    "Artist": Object {
      "albums": Array [],
      "followers": 0,
      "name": "Unknown",
    },
  },
  "News": Array [],
  "User": null,
}
`;
